### PR TITLE
RPCRecHit with timing and y-coordinate information for iRPC

### PR DESCRIPTION
**Pull request by jhgoh**
This PR changes RPC local reconstruction and dataformat to include timing and y-coordinate information from the RPCDigis.
The clustering algorithm is unchanged from the standard algorithm, timing and y position are computed as average of digis clustered into a RecHit, only when the digi contains valid timing/y info.

There are DataFormat changes to include RecHit average timing and its error (2 float variables and setter/getter functions).

I don't expect physics change with this PR, since it is addition of information and no change in the algorithm itself.

(modified)
Presented at the [RPC upgrade meeting in July 19](https://indico.cern.ch/event/557713/contributions/2248040/attachments/1311545/1962762/RPC_phase2_July_19_TJ.pdf) by Taejeong, 
and [RPC upgrade workshop in July 22](https://indico.cern.ch/event/558264/contributions/2251307/attachments/1313942/1967511/simulation_summary_RPCUpgradeWrokshop_v1.pdf) by Boris.

Uncertainty for the iRPC rolls are left to be tuned later at this moment - taking the RMS for the timing error, and distance to the boundary times 2/sqrt(12) for y coordinate error as a maximum estimate of error.

@bapavlov @mileva @monttj

### CODE REVIEW

File: RecoLocalMuon/RPCRecHit/src/RPCCluster.cc [hunk 1]
```diff
@@ -1,71 +1,84 @@
...
 }
 
-RPCCluster::~RPCCluster()
-{
-}
+RPCCluster::~RPCCluster() {}
+
+int RPCCluster::firstStrip() const { return fstrip; }
+int RPCCluster::lastStrip() const { return lstrip; }
+int RPCCluster::clusterSize() const { return -(fstrip-lstrip)+1; }
```
Comment (slava77) [inline RecoLocalMuon/RPCRecHit/src/RPCCluster.cc]: why not e.g.  ```(lstrip+1)-fstrip``` ?

**Comment by slava77** [comment]
@jhgoh is the y or time simulation already present in the phase-2 RPCs?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@jhgoh 
if you presented this in an RPC meeting, please add a link.
Thank you.

**Comment by slava77** [comment]
errors are related to this PR
@jhgoh please recall that every non-trivial PR has to be tested with ```runTheMatrix.py -l limited -i all ``` before submission.

File: RecoLocalMuon/RPCRecHit/src/RPCCluster.cc (hunk 1 above)
Comment (jhgoh) [inline RecoLocalMuon/RPCRecHit/src/RPCCluster.cc]: Just because the formula have been there, I can just guess that this is for the code readability.

**Comment by jhgoh** [comment]
@slava77 to be resolved soon.

**Comment by jhgoh** [comment]
Problem resolved - crashes due to zero y local position error. Fix is to be added soon.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@cmsbuild please test
Try again in the hopes that the DAS error will go away.

**Comment by cvuosalo** [comment]
@jhgoh: Could you please confirm that the y data and precise timing is currently included in the Phase 2 RPC simulation in standard workflows 10624.0 and 11024.0?

**Comment by jhgoh** [comment]
@cvuosalo I'm starting to check now. 
At least for the local reconstruction part, it is independent from the workflow.

**Comment by jhgoh** [comment]
Time information are available in the RPCDigis produced with 10624.0
but the y positions is empty in fact. As far as I know, the y position is not crucial for the upgrade study at this moment.

canvas_1
canvas_2
canvas_3

The RecHit part had some bug which ignores time information from single-digi recHit, so I'm now updating this PR to fix the problem.

**Comment by jhgoh** [comment]
After applying the latest commit - RPCRecHit reproduces time information of digis.

canvas_1

**Comment by cvuosalo** [comment]
+1

For #15263 85ffdf1db1631450af994027af512cccb551d0f5

Including timing and y-coordinate information in Phase 2 RPC local reco.

The code changes are satisfactory. and Jenkins tests on an earlier version of this PR against baseline CMSSW_8_1_X_2016-07-29-1100 show no significant differences.
An extended test of workflow 10624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2023tilted with 100 events against baseline CMSSW_8_1_0_pre9 shows no significant differences and also shows the addition of the new timing info to RECO event content:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    230.0 ->       238.6          9      3.6   0.00     RPCDetIdRPCRecHitsOwnedRangeMap_rpcRecHits__RECO.
-------------------------------------------------------------
  1813432 ->     1813453         21             0.0     ALL BRANCHES
```
 Jenkins tests currently are failing on all PRs, so the latest small revision of this PR has not been tested by Jenkins yet, but the previous Jenkins test and the 10624.0 test are sufficient to show there are no problems with this PR.

**Comment by slava77** [comment]
@cmsbuild please test
jenkins 81X should run OK now (140.53 removed from the matrix)

**Comment by davidlange6** [comment]
+1
@jhgoh - is it possible to move the new data format additions into the derived classes where they are actually appropriate? Not sure if we looked at the memory increase from the use unused (in run2) variables?

**Pull request merged**